### PR TITLE
Odometery calibration

### PR DESCRIPTION
**Pull request by jwallace42**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | None |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of contribution in a few bullet points
Added a new behavior tree for odometry calibration. 

## Description of documentation updates required from your changes
Added description of the odometry behavior tree.

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_behaviors/plugins/drive_on_heading.hpp [hunk 1]
```diff
@@ -93,8 +93,6 @@ class DriveOnHeading : public TimedBehavior<ActionT>
   Status onCycleUpdate()
   {
     rclcpp::Duration time_remaining = end_time_ - this->steady_clock_.now();
-    RCLCPP_ERROR_STREAM(this->logger_, "Time remaining" << time_remaining.seconds());
-    RCLCPP_ERROR_STREAM(this->logger_, "Time remaining" << command_time_allowance_.seconds());
     if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
```
Comment (jwallace42) [inline nav2_behaviors/plugins/drive_on_heading.hpp:96]: I left these in the previous PR for drive on heading.

File: nav2_bt_navigator/behavior_trees/odometry_calibration.xml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<!--
+  This Behavior Tree drives in a square
+-->
+
+<root main_tree_to_execute="MainTree">
+  <BehaviorTree ID="MainTree">
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/odometry_calibration.xml:6]: Adding a repeat node (e.g. loop N times) would be nice

File: nav2_bt_navigator/behavior_trees/odometry_calibration.xml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    <Sequence name="Drive in a square">
+      <DriveOnHeading dist_to_travel="0.2" speed="0.02" time_allowance="12"/>
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/odometry_calibration.xml:8]: Might want a bit faster than 2 cm/s, how about 10cm/s?

Comment (SteveMacenski) [reply nav2_bt_navigator/behavior_trees/odometry_calibration.xml:8]: And like 2 meters of side length

File: nav2_bt_navigator/behavior_trees/odometry_calibration.xml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      <Spin spin_dist="1.57"/>
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/odometry_calibration.xml:9]: Add a few more sig figs than that, that much could actually contribute to perceived issues if repeated 100 times. At least 5 decimals would be nice. I doubt it'll execute that accurately, but at least good to know we set it to be enough where the BT wouldn't be the cause of any issues

File: nav2_bt_navigator/behavior_trees/odometry_calibration.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<!--
+  This Behavior Tree drives in a square
```
Comment (SteveMacenski) [inline nav2_bt_navigator/behavior_trees/odometry_calibration.xml:2]:
```suggestion
  This Behavior Tree drives in a square for odometry calibration experiments
```

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
You have some linting and test failures

**Comment by SteveMacenski** [comment]
```
/opt/overlay_ws/src/navigation2/nav2_system_tests/src/behavior_tree/test_behavior_tree_node.cpp:213
Expected equality of these values:
  bt_handler->loadBehaviorTree(entry.path().string())
    Which is: false
  true
```

it looksl ike your BT is somehow invalid?

**Comment by jwallace42** [comment]
So I fixed the loading issue but I ran into another error that is quite confusing. 

That test is failing due to a timeout. On further inspection I found the line that was causing the issue. 

https://github.com/jwallace42/navigation2/blob/main/nav2_system_tests/src/behavior_tree/test_behavior_tree_node.cpp#L141. 

The test only times out when the DriveOnHeading BT is in the new xml file. 

This is quite odd for the following reasons. 

1. We have no trouble creating the tree when running the full stack. 
2. The test for DriveOnHeading does not time out. 

I would expect the issue to  time out every time the BT was trying to load DriveOnHeading. I am not quite sure what is going on. It seems to be deeper with the behavior tree library.

**Comment by SteveMacenski** [comment]
Is it because its not in the library's list? https://github.com/jwallace42/navigation2/blob/main/nav2_system_tests/src/behavior_tree/test_behavior_tree_node.cpp#L54-L92

I think that's it, if it doesn't have access to the library, then it can't load. But I'm floored that BT.CPP doesn't throw an error and just hangs instead :eye roll:

**Comment by jwallace42** [comment]
I wish that was the issue but I have it here https://github.com/ros-planning/navigation2/pull/2934/files.

**Comment by SteveMacenski** [comment]
Huh, that's super strange

**Comment by jwallace42** [comment]
I dug into this a bit more. 

The new Cancel nodes have the same issue. 


I tried digging into test with gdb but came across another unexpected result. 
The test with pass with colcon --packages-select nav2_system_tests but fail if I try to run the executable directly. 

`>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'Service type support not from this implementation. Got:
    Handle's typesupport identifier (rosidl_typesupport_cpp) is not supported by this library, at /tmp/binarydeb/ros-rolling-rosidl-typesupport-cpp-1.4.2/src/type_support_dispatch.hpp:111
    Could not load library libnav2_msgs__rosidl_typesupport_introspection_cpp.so: dlopen error: libnav2_msgs__rosidl_typesupport_introspection_cpp.so: cannot open shared object file: No such file or directory, at /tmp/binarydeb/ros-rolling-rcutils-5.0.1/src/shared_library.c:99, at /tmp/binarydeb/ros-rolling-rosidl-typesupport-cpp-1.4.2/src/type_support_dispatch.hpp:76
while fetching it, at /tmp/binarydeb/ros-rolling-rmw-cyclonedds-cpp-1.1.2/src/rmw_node.cpp:4174'
`

**Comment by SteveMacenski** [comment]
Neat... I wonder if something about the messages wasn't exported right for the drive on heading work. It could be that we missed some export that is causing this since its unique to that (and wasn't even introduced in this PR)

**Comment by jwallace42** [comment]
Yeah, checked though the cmake and didn't find any discrepancies between DriveOnHeading and other plugins. 

I also see this issue in 

`CancelWait`
`CancelSpin`
'CancelBackUp'
'CancelDriveOnHeading'

**Comment by SteveMacenski** [comment]
Is it just the cancel plugins?

@padhupradheep it sounds like we may have missed something on the cancel behavior tree plugins you worked on

**Comment by padhupradheep** [comment]
I was actually planning to jump in the conversation. 

I dug deep into this during that PR. https://github.com/ros-planning/navigation2/pull/2787#issuecomment-1026115084

For some reason the node didn’t register at all..

**Comment by jwallace42** [comment]
So this is BT bug?

**Comment by jwallace42** [comment]
I was just looking at the backtrace here 

`#0  futex_abstimed_wait_cancelable (private=<optimized out>, abstime=0x7fffffffa5b0, clockid=<optimized out>, expected=0, futex_word=0x5555557bb74c) at ../sysdeps/nptl/futex-internal.h:320
#1  __pthread_cond_wait_common (abstime=0x7fffffffa5b0, clockid=<optimized out>, mutex=0x5555557bb6f8, cond=0x5555557bb720) at pthread_cond_wait.c:520
#2  __pthread_cond_timedwait (cond=0x5555557bb720, mutex=0x5555557bb6f8, abstime=0x7fffffffa5b0) at pthread_cond_wait.c:656
#3  0x00007ffff7e051ce in rclcpp::node_interfaces::NodeGraph::wait_for_graph_change(std::shared_ptr<rclcpp::Event>, std::chrono::duration<long, std::ratio<1l, 1000000000l> >) ()
   from /opt/ros/rolling/lib/librclcpp.so
#4  0x00007ffff7f78c67 in rclcpp_action::ClientBase::wait_for_action_server_nanoseconds(std::chrono::duration<long, std::ratio<1l, 1000000000l> >) () from /opt/ros/rolling/lib/librclcpp_action.so
#5  0x00007fffed93a8a9 in nav2_behavior_tree::BtCancelActionNode<nav2_msgs::action::DriveOnHeading>::BtCancelActionNode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&) ()
   from /home/josh/nav2_ws/install/nav2_behavior_tree/lib/libnav2_drive_on_heading_cancel_bt_node.so
#6  0x00007fffed92be4d in nav2_behavior_tree::DriveOnHeadingCancel::DriveOnHeadingCancel(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&) () from /home/josh/nav2_ws/install/nav2_behavior_tree/lib/libnav2_drive_on_heading_cancel_bt_node.so
#7  0x00007fffed92befd in std::_Function_handler<std::unique_ptr<BT::TreeNode, std::default_delete<BT::TreeNode> > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&), BT_RegisterNodesFromPlugin::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&)#1}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&) ()
   from /home/josh/nav2_ws/install/nav2_behavior_tree/lib/libnav2_drive_on_heading_cancel_bt_node.so
#8  0x00007ffff7afeeef in BT::BehaviorTreeFactory::instantiateTreeNode(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::NodeConfiguration const&) const () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#9  0x00007ffff7b1e77b in ?? () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#10 0x00007ffff7b1f3d1 in ?? () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#11 0x00007ffff7b1f670 in ?? () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#12 0x00007ffff7b1f670 in ?? () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#13 0x00007ffff7b1ef9c in BT::XMLParser::Pimpl::recursivelyCreateTree(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, BT::Tree&, std::shared_ptr<BT::Blackboard>, std::shared_ptr<BT::TreeNode> const&) () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#14 0x00007ffff7b1f1ac in BT::XMLParser::instantiateTree(std::shared_ptr<BT::Blackboard> const&) () from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#15 0x00007ffff7b02fd3 in BT::BehaviorTreeFactory::createTreeFromText(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<BT::Blackboard>) ()
   from /opt/ros/rolling/lib/libbehaviortree_cpp_v3.so
#16 0x00005555555ec6e3 in BehaviorTreeHandler::loadBehaviorTree (this=0x5555557b60a0, filename="/home/josh/nav2_ws/install/nav2_bt_navigator/share/nav2_bt_navigator/behavior_trees/odometry_calibration.xml")
    at /usr/include/c++/9/ext/atomicity.h:96
#17 0x00005555555be5dc in BehaviorTreeTestFixture_TestBTXMLFiles_Test::TestBody (this=<optimized out>) at /usr/include/boost/filesystem/path.hpp:454
#18 0x0000555555658d91 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (location=0x55555566a672 "the test body", method=<optimized out>, object=0x55555571f9d0)
    at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2414
#19 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x55555571f9d0, method=<optimized out>, location=location@entry=0x55555566a672 "the test body")
    at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2469
#20 0x000055555564cff6 in testing::Test::Run (this=0x55555571f9d0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2508
#21 testing::Test::Run (this=0x55555571f9d0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2498
#22 0x000055555564d155 in testing::TestInfo::Run (this=0x5555556abfc0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2684
#23 testing::TestInfo::Run (this=0x5555556abfc0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2657
#24 0x000055555564d27d in testing::TestSuite::Run (this=0x5555556ac5a0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2816
#25 testing::TestSuite::Run (this=0x5555556ac5a0) at /opt/ros/rolling/src/gtest_vendor/./src/gtest.cc:2795
#26 0x000055555564d823 in testing::internal::UnitTestImpl::RunAllTests (this=0x5555556ac170) at /usr/include/c++/9/bits/stl_vector.h:1040
#27 0x000055555564da88 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (location=0x55555566ba28 "auxiliary test code (environments or event listeners)", `

**Comment by SteveMacenski** [comment]
I see no reason why it should matter, but none of those have `on_tick`'s or ports in the provided port functions.

**Comment by padhupradheep** [comment]
hey @jwallace42, if you need some help here.. I can give you an hand here.. provided I know, some more details from your last update..

**Comment by jwallace42** [comment]
@padhupradheep not really sure how to go about solving the issue as a result I hadn't done anything this past week.

If you want to reproduce just put one of the troublesome BT nodes in the behavior tree xml and run the nav2_system_test. It should fail due to a timeout.

**Comment by padhupradheep** [comment]
let me give it a dig and come back..

**Comment by jwallace42** [comment]
Cool! Cool! Let me know if you have any questions.... 

I tried a lot of different things :)

**Comment by SteveMacenski** [comment]
@padhupradheep find anything?

**Comment by padhupradheep** [comment]
not yet.. need few more days..

**Comment by padhupradheep** [comment]
Strangely, I was not able to reproduce the issue. I added the "so-mentioned" problematic behavior nodes to the `test_behavior_tree_node.cpp` and seems the tests ran without any issue.. 

I hope, I'm not missing something else..

**Comment by padhupradheep** [comment]
I think, there is not much that I can find here. Unfortunately!

**Comment by SteveMacenski** [comment]
What did you try? Just so we can use that as a basis of what doesn't work

**Comment by SteveMacenski** [comment]
https://github.com/BehaviorTree/BehaviorTree.CPP/issues/390 I also filed a ticket

**Comment by SteveMacenski** [comment]
@jwallace42 please rebase, I doubt it'll fix it, but we updated docker images to 22.04

**Comment by padhupradheep** [comment]
Oh wait, I see the problematic nodes are being registered after all.. 

I was doing some tests now.. I made our regular navigate_to_pose BT node into a (non-sense) BT, by adding the following changes:


```
<RecoveryNode number_of_retries="1" name="FollowPath">
            <Sequence name="CancelingControlAndWait">
                <FollowPath path="{path}" controller_id="FollowPath"/>
                <ClearEntireCostmap name="ClearLocalCostmap-Context" service_name="local_costmap/clear_entirely_local_costmap"/>
                <CancelControl name="ControlCancel"/>
                <Wait wait_duration="20"/>
                <CancelWait name="CancelWait"/>
            </Sequence>
            <ClearEntireCostmap name="ClearLocalCostmap-Context" service_name="local_costmap/clear_entirely_local_costmap"/>
 </RecoveryNode>
```

So, what I inferred is that both the `CancelControl` and `CancelWait` BT nodes were ticked.. Of course the tests given in the https://github.com/ros-planning/navigation2/blob/main/nav2_system_tests/src/behavior_tree/test_behavior_tree_node.cpp failed.. But there was actually no timeout. 

I verified the tick by just adding a print to the tick() function found in the bt_cancel_node.hpp

**Comment by SteveMacenski** [comment]
Weirder yet, its been reported as the result of a problem in the other system tests: https://github.com/ros-planning/navigation2/pull/2962#issuecomment-1139798508

Something seems wrong with that test or BT.CPP...

**Comment by SteveMacenski** [comment]
My comment in https://github.com/ros-planning/navigation2/pull/2962#issuecomment-1139890269 explains why we might be seeing it in that particular PR. It might be worth adding some logging inside of the BT action server to see where and what node is hanging, and specifically where.

**Comment by SteveMacenski** [comment]
Rebase or pull in main, there are some failing tests which concern me but I believe they've been handled elsewhere

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
CI failed in an unusual way, retriggering

**Pull request merged**